cabal-install/src/Distribution/Client/VCS.hs

```diff
@@ -305,7 +314,43 @@ vcsDarcs =
 
     vcsSyncRepos :: Verbosity -> ConfiguredProgram
                  -> [(SourceRepositoryPackage f, FilePath)] -> IO [MonitorFilePath]
-    vcsSyncRepos _v _p _rs = fail "sync repo not yet supported for darcs"
+    vcsSyncRepos verbosity prog = \ case
+      [] -> pure []
+      (primaryRepo, primaryLocalDir) : secondaryRepos ->
+        [ monitorDirectoryExistence dir
+        | dir <- primaryLocalDir : (snd <$> secondaryRepos) ] <$ do
+        vcsSyncRepo verbosity prog primaryRepo primaryLocalDir Nothing
+        sequence_
+          [ vcsSyncRepo verbosity prog repo localDir (Just primaryLocalDir)
+          | (repo, localDir) <- secondaryRepos ]
+
+    vcsSyncRepo verbosity prog SourceRepositoryPackage{..} localDir _peer =
+      try (lines <$> darcsWithOutput localDir ["log", "--last", "1"]) >>= \ case
+        Right (_:_:_:x:_)
+          | Just tag <- (List.stripPrefix "tagged " . List.dropWhile Char.isSpace) x
+          , Just tag' <- srpTag
+          , tag == tag' -> pure ()
+        Left (_ :: SomeException) -> pure ()
+        _ -> do
+          removeDirectoryRecursive localDir `catch` \ case
+              e | isDoesNotExistError e -> pure ()
+                | otherwise -> throw e
+          darcs (takeDirectory localDir) cloneArgs
+      where
+        darcs :: FilePath -> [String] -> IO ()
+        darcs = darcs' runProgramInvocation
+
+        darcsWithOutput :: FilePath -> [String] -> IO String
+        darcsWithOutput = darcs' getProgramInvocationOutput
+
+        darcs' f cwd args = f verbosity (programInvocation prog args)
+          { progInvokeCwd = Just cwd }
+
+        cloneArgs = ["clone"] <|> tagArgs <|> [srpLocation, localDir] <|> verboseArg
```

Comment:
Similarly to `pure`/`return`, other similar code uses `++` instead of `<|>`. Same below.

Reply:
Done